### ISSUE TITLE
Repeated sequence_index values in specific situations

### ISSUE THREAD

**Issue by Ng-ms** [sdv-dev/SDV#2004]
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* SDV version: 1.12.0
* Python version: 3.12
* Operating System: LInux 

### 
after applying the workaround in #1973 , the sequnece_index is not Nan anymore, but I notice a repeated sequence index (In my dataset is the date of the visit )  so this date should be unique because it is a unique visit, 
![Screenshot from 2024-05-14 14-42-07](https://github.com/sdv-dev/SDV/assets/17408097/a322d733-f370-48b3-b871-a16d9bc713eb)

**Comment by npatki** [comment]
Thanks for filing @Ng-ms. Adding a link to [our previous conversation](https://github.com/sdv-dev/SDV/issues/1973#issuecomment-2110128541).

As with all these issues, it is very beneficial to us if you can provide some data to help us replicate it. Otherwise, it may take us longer to find the root cause.

### Summary
- The PARSynthesizer should be creating data that is similar to real data. If real data has repeating sequence indices, then we would also expect the synthetic data to have sequence indices
- You have confirmed that in this case, the real data does _not_ have repeating sequence indices.

### Next Steps
The SDV team will investigate the cause of this issue. Internally, I would want to check for two things:

1. We compute the differences in the sequence index columns, and then later synthesize them. When we do this, we enforce min/max values -- meaning that we should never synthesize any interval that is <min and >max, as observed from the real data. We should verify this is working.
2. Perhaps what we're seeing then could be a rounding error? Eg instead of synthesizing +1 day we might be synthesizing +23hr 59min, which is being rounded down for some reason

**Comment by Ng-ms** [comment]
Thank you, @npatki, unfortunately, I am not able to share the data since it belongs to a hospital, 

``
min_max_scaler = MinMaxScaler()
df[numeric_columns] = min_max_scaler.fit_transform(df[numeric_columns])
df[date_columns] = df[date_columns].apply(pd.to_datetime,format='%d/%m/%Y', errors= 'coerce') 
df['pre_date'] = pd.to_datetime(df['re_date'],  unit= 'ns').astype(int)
metadata.set_sequence_index(column_name='visit_date')
synthesizer = PARSynthesizer(metadata,epochs=1000, context_columns= ['pre_date',sex,'Cod',], verbose=True,  enforce_min_max_values=True, enforce_rounding=True,  cuda=True)
synthesizer.fit(df)
synthetic_data = synthesizer.sample(num_sequences=4000,sequence_length=None)
... 
this is the code I am using right now

**Comment by Ng-ms** [comment]
Thank you, @npatki, unfortunately, I am not able to share the data since it belongs to a hospital,

`     min_max_scaler = MinMaxScaler()
df[numeric_columns] = min_max_scaler.fit_transform(df[numeric_columns])
df[date_columns] = df[date_columns].apply(pd.to_datetime,format='%d/%m/%Y', errors= 'coerce')
df['pre_date'] = pd.to_datetime(df['re_date'], unit= 'ns').astype(int)
metadata.set_sequence_index(column_name='visit_date')
synthesizer = PARSynthesizer(metadata,epochs=1000, context_columns= ['pre_date',sex,'Cod',], verbose=True, enforce_min_max_values=True, enforce_rounding=True, cuda=True)
synthesizer.fit(df)
synthetic_data = synthesizer.sample(num_sequences=4000,sequence_length=None) `     
this is the code I am using right now

**Comment by Ng-ms** [comment]
![image](https://github.com/sdv-dev/SDV/assets/17408097/a674b221-ddf0-4e26-8f88-ccc73603f9d6)

**Comment by Ng-ms** [comment]
Thank you, @npatki, unfortunately, I am not able to share the data since it belongs to a hospital,

`min_max_scaler = MinMaxScaler()
df[numeric_columns] = min_max_scaler.fit_transform(df[numeric_columns])
df[date_columns] = df[date_columns].apply(pd.to_datetime,format='%d/%m/%Y', errors= 'coerce')
df['pre_date'] = pd.to_datetime(df['pre_date'], unit= 'ns').astype(int)
metadata.set_sequence_index(column_name='visit_date')
synthesizer = PARSynthesizer(metadata,epochs=1000, context_columns= ['pre_date',sex,'Cod',], verbose=True, enforce_min_max_values=True, enforce_rounding=True, cuda=True)
synthesizer.fit(df)
synthetic_data = synthesizer.sample(num_sequences=4000,sequence_length=None) ` 
this is the code I am using right now . 
 in the image below you can see the number of duplicated dates of   "visit_date " for the synthetic data 
![image](https://github.com/sdv-dev/SDV/assets/17408097/f908232f-d0c3-4514-9b52-af5478ff6fe5)

while here in the real data,   Series([], dtype: int64) ,

**Comment by Ng-ms** [comment]
hello, yes  pre_date is a context column and visit_date is a sequence index

**Comment by srinify** [comment]
Hi there @Ng-ms I tried to replicate this with our demo datasets and wasn't able to yet unfortunately.

If you're able to meet us halfway by trying to modify one of our demo datasets (deleting rows, adding context columns, formatting date time values in a similar way to your dataset, etc) to force this error to occur, that would be a massive help for us.

The following code snippet can get you started with a demo dataset that has a datetime column:

```
from sdv.datasets.demo import download_demo

real_data, metadata = download_demo(
    modality='sequential',
    dataset_name='nasdaq100_2019'
)
```

Either way, we'll keep this issue open to see if others have run into it and can help us reproduce this.

**Comment by srinify** [comment]
I was able to reproduce this behavior! Sharing code now:

```
import pandas as pd
import numpy as np
from sdv.sequential import PARSynthesizer
from sdv.metadata.single_table import SingleTableMetadata

np.random.seed(1)
num_sequences = 10
rows_per_sequence = 10
total_rows = num_sequences * rows_per_sequence

# Random Dates
start_date, end_date = pd.to_datetime('2020-01-01'), pd.to_datetime('2024-05-01')
random_dates = pd.to_datetime(np.random.randint(start_date.value, end_date.value, size=total_rows))

# Context Column Dates
context_date_column = []
for _ in range(num_sequences):
  random_date = pd.to_datetime(np.random.randint(start_date.value, end_date.value))
  sequence = [random_date] * rows_per_sequence
  context_date_column += sequence

# Row ID
id = [i for i in range(0, total_rows)]

# Sequence Key
values = np.repeat(np.arange(10), 10)
values_list = values.tolist()
s_key = values_list

# Sex
sex = ["Male" for i in range(0, total_rows)]

df3 = pd.DataFrame( {
    'id': id,
    's_key': s_key,
    'visits': random_dates,
    'pre_date': context_date_column,
    'sex': sex
})

df3['visits'] = df3['visits'].dt.date
df3['pre_date'] = df3['pre_date'].dt.date

# Check for Duplicate Sequence Index Values in Source Data
def check_duplicate_seq_index(df, seq_index_column):
  for i in df['s_key'].unique():
    seq_df = df[df['s_key'] == i]
    has_duplicates = seq_df[seq_index_column].duplicated().any()
    if has_duplicates:
      return True
    return False

check_duplicate_seq_index(df3, 'visits')

from sklearn.preprocessing import MinMaxScaler

min_max_scaler = MinMaxScaler()

# Datetime Columns
df3[['visits', 'pre_date']] = df3[['visits', 'pre_date']].apply(pd.to_datetime, format='%Y-%m-%d', errors='coerce')
df3['pre_date'] = pd.to_datetime(df3['pre_date'], unit='ns').astype(int)

from sdv.metadata import SingleTableMetadata

metadata3 = SingleTableMetadata()
metadata3.detect_from_dataframe(df3)
metadata3.update_column(column_name='pre_date', sdtype='numerical')
metadata3.update_column(column_name='s_key', sdtype='id')
metadata3.set_sequence_key('s_key')
metadata3.set_sequence_index('visits')
synth3 = PARSynthesizer(metadata3, epochs=1000, context_columns=['pre_date', 'sex'], verbose=True, enforce_min_max_values=True, enforce_rounding=True)
synth3.fit(df3)
synth_df3 = synth3.sample(num_sequences=10,sequence_length=None)

check_duplicate_seq_index(synth_df3, 'visits')
```

Final line returns `True`

**Comment by srinify** [comment]
@Scit3ch very thorough & excellent analysis! You seem to be right about this and I was able to replicate this on a much simpler dataset where I experimented with decreasing `sequence_length` to approach the number of unique values in the original number of unique values in a single sequence in the original dataset. 

In simpler terms, trying to synthesize a sequence of 7 rows where one of the original sequences only had 5 rows caused SDV to generate duplicates and essentially "run out" of dates to generate because of `enforce_min_max_values` being True.

Context columns don't seem to matter here. Thanks for all the help, I've opened this bug report: https://github.com/sdv-dev/SDV/issues/2031

Some short term workarounds with tradeoffs:

- You can keep `enforce_min_max_values` as False and this will remove the max value ceiling for the datetime sequence_ ey column. But this will mean that the synthesized data will be less representative of your real data so this is a big tradeoff until this bug is fixed.
- You can set `num_sequences` to be identical to the number of rows in your smallest, least unique (when it comes to the sequence key column) sequence from your real data. E.g. if you have a small sequence with 5 unique values for the sequence key, don't generate more than 5 rows per sequence. But this is also a limitation of SDV until this bug is fixed.
